pkg/sink/codec/bootstraper.go

```diff
@@ -0,0 +1,263 @@
+		return nil
+	}
+	table.reset()
+	tableInfo := table.tableInfo.Load().(*model.TableInfo)
+	events, err := b.generateEvents(table.topic, table.totalPartition.Load(), tableInfo)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for _, event := range events {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case b.outCh <- event:
+		}
+	}
+	return nil
+}
+
+func (b *bootstrapWorker) generateEvents(
+	topic string,
+	totalPartition int32,
+	tableInfo *model.TableInfo,
+) ([]*future, error) {
+	res := make([]*future, 0, int(totalPartition))
+	// Bootstrap messages of a table should be sent to all partitions.
+	for i := 0; i < int(totalPartition); i++ {
+		encoder := b.builder.Build()
+		msg, err := encoder.EncodeDDLEvent(&model.DDLEvent{
+			StartTs:     0,
+			CommitTs:    0,
+			TableInfo:   tableInfo,
+			IsBootstrap: true,
+		})
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		key := TopicPartitionKey{
+			Topic:          topic,
+			Partition:      int32(i),
+			TotalPartition: totalPartition,
+		}
+
+		future := &future{
+			Key:  key,
+			done: make(chan struct{}),
+		}
+		future.Messages = append(future.Messages, msg)
+		close(future.done)
+		res = append(res, future)
+	}
+	return res, nil
+}
+
+func (b *bootstrapWorker) gcInactiveTables() {
+	b.activeTables.Range(func(key, value interface{}) bool {
+		table := value.(*tableStatus)
+		if !table.isActive(b.maxInactiveDuration) {
+			log.Info("A table is removed from the bootstrap worker",
+				zap.Int64("tableID", table.id),
+				zap.String("topic", table.topic),
+				zap.Stringer("changefeed", b.changefeedID))
+			b.activeTables.Delete(key)
+		}
+		return true
+	})
+}
+
+// tableStatus is used to record the status of a table
+type tableStatus struct {
+	// id is the table's ID, it will not change
+	id int64
+	// topic is the table's topic, it will not change
+	topic string
+	// All fields below are concurrently accessed, please use atomic operations.
+	// totalPartition is the total number of partitions of the table's topic
+	totalPartition atomic.Int32
+	// counter is the number of row event sent since last bootstrap message sent
+	// It is used to check if the bootstrap message should be sent
+	counter atomic.Int32
+	// lastMsgReceivedTime is the last time the row event is received
+	// It is used to check if the table is inactive
+	lastMsgReceivedTime atomic.Value
+	// lastSendTime is the last time the bootstrap message is sent
+	// It is used to check if the bootstrap message should be sent
+	lastSendTime atomic.Value
+	// version is the table version
+	// It is used to check if the table schema is changed since last bootstrap message sent
+	version atomic.Uint64
+	// tableInfo is the tableInfo of the table
+	// It is used to generate bootstrap message
+	tableInfo atomic.Value
+}
+
+func newTableStatus(key TopicPartitionKey, row *model.RowChangedEvent) *tableStatus {
+	res := &tableStatus{
+		id:    row.Table.TableID,
+		topic: key.Topic,
+	}
+	res.counter.Add(1)
```

Comment:
Why the count is set `1`

Reply:
A `tableStatistic` will only be created when a `rowChangeEvent` goes through the `bootstrapWorker`, so the counter should add `1`.